File: src/engine/events.js
// Crisis & Event System
// Ported from sim/events.py

const CRISIS_TEMPLATES = [
  { key: "financial_panic", name: "Financial Panic", severity: 0.7, duration: 2, gdpImpact: -0.08, instImpact: -0.05, socialImpact: -0.10, dImpact: -3, reformBonus: 2.5, triggers: ["political_instability", "sovereign_debt_crisis"] },
  { key: "sovereign_debt_crisis", name: "Sovereign Debt Crisis", severity: 0.75, duration: 3, gdpImpact: -0.12, instImpact: -0.08, socialImpact: -0.15, dImpact: -5, reformBonus: 3.0, triggers: ["political_instability", "social_unrest"] },
  { key: "hyperinflation", name: "Hyperinflation", severity: 0.85, duration: 2, gdpImpact: -0.15, instImpact: -0.10, socialImpact: -0.20, dImpact: -8, reformBonus: 3.5, triggers: ["political_instability", "social_unrest", "regime_change"] },
  { key: "political_instability", name: "Political Instability", severity: 0.6, duration: 1, gdpImpact: -0.03, instImpact: -0.12, socialImpact: -0.10, dImpact: -2, reformBonus: 4.0, triggers: ["financial_panic", "ethnic_conflict", "regime_change"] },
  { key: "ethnic_conflict", name: "Ethnic Conflict Flare-Up", severity: 0.8, duration: 3, gdpImpact: -0.10, instImpact: -0.15, socialImpact: -0.30, dImpact: -6, reformBonus: 1.5, triggers: ["political_instability", "external_intervention"] },
  { key: "environmental_disaster", name: "Environmental Disaster", severity: 0.5, duration: 1, gdpImpact: -0.05, instImpact: -0.02, socialImpact: -0.05, dImpact: -2, reformBonus: 2.0, triggers: ["social_unrest"] },
  { key: "commodity_price_collapse", name: "Commodity Price Collapse", severity: 0.6, duration: 2, gdpImpact: -0.07, instImpact: -0.03, socialImpact: -0.08, dImpact: -1, reformBonus: 2.5, triggers: ["sovereign_debt_crisis", "political_instability"] },
  { key: "external_intervention", name: "External Intervention", severity: 0.9, duration: 4, gdpImpact: -0.15, instImpact: -0.20, socialImpact: -0.25, dImpact: -10, reformBonus: 2.0, triggers: ["political_instability", "ethnic_conflict", "regime_change"] },
  { key: "demographic_shock", name: "Demographic Shock", severity: 0.4, duration: 5, gdpImpact: -0.04, instImpact: -0.02, socialImpact: -0.10, dImpact: -1, reformBonus: 1.8, triggers: ["social_unrest"] },
  { key: "pandemic", name: "Pandemic", severity: 0.7, duration: 2, gdpImpact: -0.06, instImpact: -0.03, socialImpact: -0.12, dImpact: -2, reformBonus: 2.8, triggers: ["financial_panic", "political_instability"] },
  { key: "social_unrest", name: "Social Unrest", severity: 0.5, duration: 1, gdpImpact: -0.04, instImpact: -0.05, socialImpact: -0.10, dImpact: -2, reformBonus: 3.0, triggers: ["political_instability"] },
  { key: "regime_change", name: "Regime Change", severity: 0.9, duration: 3, gdpImpact: -0.10, instImpact: -0.20, socialImpact: -0.15, dImpact: -8, reformBonus: 5.0, triggers: ["political_instability", "ethnic_conflict", "social_unrest"] },
];

const EVENT_TEMPLATES = [
  { key: "technology_breakthrough", name: "Technology Breakthrough", prob: 0.03, gdpImpact: 0.02, dChange: 2 },
  { key: "foreign_investment_wave", name: "Foreign Investment Wave", prob: 0.04, gdpImpact: 0.03, dChange: 1.5, instChange: 0.02 },
  { key: "diaspora_return", name: "Diaspora Return Wave", prob: 0.03, gdpImpact: 0.02, dChange: 3, instChange: 0.03 },
  { key: "resource_discovery", name: "New Resource Discovery", prob: 0.02, gdpImpact: 0.04, dChange: -1, instChange: -0.02 },
  { key: "cultural_renaissance", name: "Cultural Renaissance", prob: 0.04, dChange: 2, instChange: 0.01, socialChange: 0.05 },
  { key: "trade_agreement", name: "Favorable Trade Agreement", prob: 0.05, gdpImpact: 0.02, dChange: 1 },
  { key: "leader_emergence", name: "Transformational Leader", prob: 0.02, dChange: 1, instChange: 0.05, socialChange: 0.05 },
];

export class EventEngine {
  constructor(seed = null) {
    this.seed = seed || Math.floor(Math.random() * 1000000);
    this.rng = this._createRng(this.seed);
    this.activeCrises = [];
    this.crisisHistory = [];
    this.eventHistory = [];
  }

  _createRng(s) {
    // Simple mulberry32 PRNG
    let state = s;
    return () => {
      state |= 0; state = state + 0x6D2B79F5 | 0;
      var t = Math.imul(state ^ state >>> 15, 1 | state);
      t = t + Math.imul(t ^ t >>> 7, 61 | t) ^ t;
      return ((t ^ t >>> 14) >>> 0) / 4294967296;
    };
  }

  checkCrisis(state, year, era) {
    const baseProb = 0.015;
    const modifiers = {};
    if (state.institutionalQuality < 0.3) modifiers.political_instability = 2.5;
    else if (state.institutionalQuality < 0.5) modifiers.political_instability = 1.5;
    if (state.ethnicFragmentationIndex > 0.6) modifiers.ethnic_conflict = 2.0;
    else if (state.ethnicFragmentationIndex > 0.4) modifiers.ethnic_conflict = 1.3;
    if (state.ethnicTensionStored > 0.5) modifiers.ethnic_conflict = (modifiers.ethnic_conflict || 1) + 2;
    if (state.resourceDependency > 0.4) modifiers.commodity_price_collapse = 2.5;
    if (state.fiscalHealth < 0.3) modifiers.sovereign_debt_crisis = 2.0;
    if (state.environmentalDamageAccumulated > 0.6) modifiers.environmental_disaster = 3.0;

    for (const ct of CRISIS_TEMPLATES) {
      let prob = baseProb * (modifiers[ct.key] || 1);
      if (era === "era1_nation_building") {
        if (["hyperinflation", "pandemic"].includes(ct.key)) prob *= 0.5;
        if (ct.key === "external_intervention") prob *= 2.0;
      } else if (era === "era2_planned_development") {
        if (ct.key === "financial_panic") prob *= 0.3;
        if (ct.key === "environmental_disaster") prob *= 1.5;
      } else if (era === "era3_convergence") {
        if (ct.key === "financial_panic") prob *= 2.0;
        if (ct.key === "sovereign_debt_crisis") prob *= 1.5;
      }
      if (this.rng() < prob) {
        const crisis = { ...ct, remainingDuration: ct.duration };
        this.activeCrises.push(crisis);
        this.crisisHistory.push({ year, crisis: ct.key, severity: ct.severity });
        return crisis;
      }
    }
    return null;
  }

  checkRandomEvent() {
    for (const et of EVENT_TEMPLATES) {
      if (this.rng() < et.prob) {
        this.eventHistory.push({ event: et.key });
        return { ...et };
      }
    }
    return null;
  }

  cascadeCheck() {
    const newCrises = [];
    for (const active of this.activeCrises) {
      for (const triggerKey of (active.triggers || [])) {
        const template = CRISIS_TEMPLATES.find(c => c.key === triggerKey);
        if (!template) continue;
        if (this.rng() < 0.12) {
          if (!this.activeCrises.find(c => c.key === triggerKey)) {
            const cascade = { ...template, remainingDuration: template.duration };
            this.activeCrises.push(cascade);
            newCrises.push(cascade);
          }
        }
      }
    }
    return newCrises;
  }

  resolveCrises() {
    const resolved = [];
    this.activeCrises = this.activeCrises.filter(c => {
      c.remainingDuration--;
      if (c.remainingDuration <= 0) { resolved.push(c); return false; }
      return true;
    });
    return resolved;
  }

  get reformMultiplier() {
    if (!this.activeCrises.length) return 1.0;
    const totalBonus = this.activeCrises.reduce((s, c) => s + c.reformBonus, 0);
    return 1.0 + Math.sqrt(totalBonus) * 0.5;
  }

  get totalImpact() {
    const impact = { gdp: 0, inst: 0, social: 0, d: 0 };
    for (const c of this.activeCrises) {
      impact.gdp += c.gdpImpact;
      impact.inst += c.instImpact;
      impact.social += c.socialImpact;
      impact.d += c.dImpact;
    }
    return impact;
  }
}

File: sim/events.py
"""
Crisis & Event System

Crises are not just bad events — they are the only times when discontinuous
institutional change is possible. During normal times, reform is incremental
(1-3% per year). During a crisis, players can push through 15-30% changes.

Crisis types cascade: a financial crisis can trigger political crisis;
political crisis can trigger ethnic conflict. Smart play involves preparing
for crises (building buffers, diversifying, investing in social cohesion).
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Callable
import random
import math
from enum import Enum


class CrisisCategory(Enum):
    ECONOMIC = "economic"
    POLITICAL = "political"
    SOCIAL = "social"
    ENVIRONMENTAL = "environmental"
    EXTERNAL = "external"


@dataclass
class Crisis:
    """A crisis event that opens a reform window."""
    name: str
    category: CrisisCategory
    severity: float  # 0-1
    description: str
    reform_window_bonus: float  # multiplier on reform effectiveness during crisis
    duration_years: int = 1
    
    # Effects on nation state
    gdp_impact_pct: float = 0.0  # e.g., -0.05 = 5% GDP drop
    institutional_quality_impact: float = 0.0
    social_cohesion_impact: float = 0.0
    d_coefficient_impact: float = 0.0
    
    # Cascade triggers
    can_trigger: List[str] = field(default_factory=list)


@dataclass
class Event:
    """A non-crisis event — beneficial, neutral, or mildly negative."""
    name: str
    description: str
    probability_per_year: float = 0.05
    
    # Effects
    gdp_impact_pct: float = 0.0
    d_coefficient_change: float = 0.0
    institutional_quality_change: float = 0.0
    social_cohesion_change: float = 0.0
    population_change_pct: float = 0.0


# Crisis definitions
CRISIS_TEMPLATES = {
    "financial_panic": Crisis(
        name="Financial Panic",
        category=CrisisCategory.ECONOMIC,
        severity=0.7,
        description="Banking system loses confidence. Credit freezes. Depositors rush to withdraw.",
        reform_window_bonus=2.5,
        duration_years=2,
        gdp_impact_pct=-0.08,
        institutional_quality_impact=-0.05,
        social_cohesion_impact=-0.10,
        d_coefficient_impact=-3.0,
        can_trigger=["political_instability", "sovereign_debt_crisis"],
    ),
    "sovereign_debt_crisis": Crisis(
        name="Sovereign Debt Crisis",
        category=CrisisCategory.ECONOMIC,
        severity=0.75,
        description="Government cannot service its debt. Default looms. International creditors demand austerity.",
        reform_window_bonus=3.0,
        duration_years=3,
        gdp_impact_pct=-0.12,
        institutional_quality_impact=-0.08,
        social_cohesion_impact=-0.15,
        d_coefficient_impact=-5.0,
        can_trigger=["political_instability", "social_unrest"],
    ),
    "hyperinflation": Crisis(
        name="Hyperinflation",
        category=CrisisCategory.ECONOMIC,
        severity=0.85,
        description="Currency collapses. Prices double weekly. Savings evaporate.",
        reform_window_bonus=3.5,
        duration_years=2,
        gdp_impact_pct=-0.15,
        institutional_quality_impact=-0.10,
        social_cohesion_impact=-0.20,
        d_coefficient_impact=-8.0,
        can_trigger=["political_instability", "social_unrest", "regime_change"],
    ),
    "political_instability": Crisis(
        name="Political Instability",
        category=CrisisCategory.POLITICAL,
        severity=0.6,
        description="Government collapses or loses legitimacy. Policy paralysis. Street protests.",
        reform_window_bonus=4.0,
        duration_years=1,
        gdp_impact_pct=-0.03,
        institutional_quality_impact=-0.12,
        social_cohesion_impact=-0.10,
        d_coefficient_impact=-2.0,
        can_trigger=["financial_panic", "ethnic_conflict", "regime_change"],
    ),
    "ethnic_conflict": Crisis(
        name="Ethnic Conflict Flare-Up",
        category=CrisisCategory.SOCIAL,
        severity=0.8,
        description="Long-suppressed ethnic tensions erupt. Violence, displacement, possible separatist movements.",
        reform_window_bonus=1.5,  # Hard to reform during conflict
        duration_years=3,
        gdp_impact_pct=-0.10,
        institutional_quality_impact=-0.15,
        social_cohesion_impact=-0.30,
        d_coefficient_impact=-6.0,
        can_trigger=["political_instability", "external_intervention"],
    ),
    "environmental_disaster": Crisis(
        name="Environmental Disaster",
        category=CrisisCategory.ENVIRONMENTAL,
        severity=0.5,
        description="Major flood, drought, or industrial accident. Infrastructure damaged. Agricultural output collapses.",
        reform_window_bonus=2.0,
        duration_years=1,
        gdp_impact_pct=-0.05,
        institutional_quality_impact=-0.02,
        social_cohesion_impact=-0.05,
        d_coefficient_impact=-2.0,
        can_trigger=["social_unrest"],
    ),
    "commodity_price_collapse": Crisis(
        name="Commodity Price Collapse",
        category=CrisisCategory.ECONOMIC,
        severity=0.6,
        description="The price of your main export commodity crashes. Foreign exchange dries up.",
        reform_window_bonus=2.5,
        duration_years=2,
        gdp_impact_pct=-0.07,
        institutional_quality_impact=-0.03,
        social_cohesion_impact=-0.08,
        d_coefficient_impact=-1.0,
        can_trigger=["sovereign_debt_crisis", "political_instability"],
    ),
    "external_intervention": Crisis(
        name="External Intervention",
        category=CrisisCategory.EXTERNAL,
        severity=0.9,
        description="A great power intervenes militarily or economically. Sovereignty compromised.",
        reform_window_bonus=2.0,
        duration_years=4,
        gdp_impact_pct=-0.15,
        institutional_quality_impact=-0.20,
        social_cohesion_impact=-0.25,
        d_coefficient_impact=-10.0,
        can_trigger=["political_instability", "ethnic_conflict", "regime_change"],
    ),
    "demographic_shock": Crisis(
        name="Demographic Shock",
        category=CrisisCategory.SOCIAL,
        severity=0.4,
        description="Sudden population change — mass emigration wave, baby bust, or mortality crisis. Labor markets disrupted.",
        reform_window_bonus=1.8,
        duration_years=5,
        gdp_impact_pct=-0.04,
        institutional_quality_impact=-0.02,
        social_cohesion_impact=-0.10,
        d_coefficient_impact=-1.0,
        can_trigger=["social_unrest"],
    ),
    "pandemic": Crisis(
        name="Pandemic",
        category=CrisisCategory.SOCIAL,
        severity=0.7,
        description="Infectious disease sweeps through the population. Death, economic disruption, and social distancing.",
        reform_window_bonus=2.8,
        duration_years=2,
        gdp_impact_pct=-0.06,
        institutional_quality_impact=-0.03,
        social_cohesion_impact=-0.12,
        d_coefficient_impact=-2.0,
        can_trigger=["financial_panic", "political_instability"],
    ),
    "social_unrest": Crisis(
        name="Social Unrest",
        category=CrisisCategory.SOCIAL,
        severity=0.5,
        description="Widespread protests, strikes, and civil disobedience. The social contract frays.",
        reform_window_bonus=3.0,
        duration_years=1,
        gdp_impact_pct=-0.04,
        institutional_quality_impact=-0.05,
        social_cohesion_impact=-0.10,
        d_coefficient_impact=-2.0,
        can_trigger=["political_instability"],
    ),
    "regime_change": Crisis(
        name="Regime Change",
        category=CrisisCategory.POLITICAL,
        severity=0.9,
        description="The entire political order is overthrown. New rules, new players, uncertain outcomes.",
        reform_window_bonus=5.0,
        duration_years=3,
        gdp_impact_pct=-0.10,
        institutional_quality_impact=-0.20,
        social_cohesion_impact=-0.15,
        d_coefficient_impact=-8.0,
        can_trigger=["political_instability", "ethnic_conflict", "social_unrest"],
    ),
}

# Positive and neutral events
EVENT_TEMPLATES = {
    "technology_breakthrough": Event(
        name="Technology Breakthrough",
        description="A domestic innovation or technology transfer creates new productive capabilities.",
        probability_per_year=0.03,
        gdp_impact_pct=0.02,
        d_coefficient_change=2.0,
    ),
    "foreign_investment_wave": Event(
        name="Foreign Investment Wave",
        description="International investors discover your market. Capital and expertise flow in.",
        probability_per_year=0.04,
        gdp_impact_pct=0.03,
        d_coefficient_change=1.5,
        institutional_quality_change=0.02,
    ),
    "diaspora_return": Event(
        name="Diaspora Return Wave",
        description="Emigrants return with capital, skills, and international networks.",
        probability_per_year=0.03,
        gdp_impact_pct=0.02,
        d_coefficient_change=3.0,
        institutional_quality_change=0.03,
    ),
    "resource_discovery": Event(
        name="New Resource Discovery",
        description="Significant new mineral or energy deposits found. Wealth — but dependency risk.",
        probability_per_year=0.02,
        gdp_impact_pct=0.04,
        d_coefficient_change=-1.0,  # resource curse
        institutional_quality_change=-0.02,  # temptation of easy money
    ),
    "cultural_renaissance": Event(
        name="Cultural Renaissance",
        description="A flourishing of arts, sciences, and public discourse. Soft power and social cohesion rise.",
        probability_per_year=0.04,
        social_cohesion_change=0.05,
        d_coefficient_change=2.0,
        institutional_quality_change=0.01,
    ),
    "trade_agreement": Event(
        name="Favorable Trade Agreement",
        description="A new trade deal opens major export markets. Economic integration deepens.",
        probability_per_year=0.05,
        gdp_impact_pct=0.02,
        d_coefficient_change=1.0,
    ),
    "leader_emergence": Event(
        name="Transformational Leader Emerges",
        description="A leader of exceptional vision and competence rises to power. Reform capacity increases.",
        probability_per_year=0.02,
        d_coefficient_change=1.0,
        institutional_quality_change=0.05,
        social_cohesion_change=0.05,
    ),
}


class EventEngine:
    """Manages random events and crisis generation/propagation."""
    
    def __init__(self, seed: Optional[int] = None):
        self.rng = random.Random(seed)
        self.active_crises: List[Crisis] = []
        self.crisis_history: List[Dict] = []
        self.event_history: List[Dict] = []
    
    def check_crisis(self, nation_state: Dict, year: int, era: str) -> Optional[Crisis]:
        """Determine if a crisis triggers this year based on nation vulnerabilities.
        
        Different eras have different crisis likelihoods:
        - Era 1: Financial panics, ethnic conflict, external intervention
        - Era 2: Commodity shocks, environmental disaster (industrial)
        - Era 3: Financial crises (market transition), sovereign debt, demographic shock
        """
        # Base crisis probability per year
        base_prob = 0.015  # ~1.5% per crisis type per year
        
        # Vulnerability modifiers
        modifiers = {}
        
        # Low institutional quality increases political crisis risk
        inst_quality = nation_state.get("institutional_quality", 0.5)
        if inst_quality < 0.3:
            modifiers["political_instability"] = 2.5
        elif inst_quality < 0.5:
            modifiers["political_instability"] = 1.5
        
        # High ethnic fragmentation increases ethnic conflict risk
        ethnic_frag = nation_state.get("ethnic_fragmentation_index", 0.3)
        if ethnic_frag > 0.6:
            modifiers["ethnic_conflict"] = 2.0
        elif ethnic_frag > 0.4:
            modifiers["ethnic_conflict"] = 1.3
        
        # Stored ethnic tension (from era transitions) increases risk
        stored_tension = nation_state.get("ethnic_tension_stored", 0.0)
        if stored_tension > 0.5:
            modifiers["ethnic_conflict"] = modifiers.get("ethnic_conflict", 1.0) + 2.0
        
        # Resource dependency increases commodity crisis risk
        resource_dependency = nation_state.get("resource_dependency", 0.0)
        if resource_dependency > 0.4:
            modifiers["commodity_price_collapse"] = 2.5
        
        # Low fiscal health increases sovereign debt crisis risk
        fiscal_health = nation_state.get("fiscal_health", 0.5)
        if fiscal_health < 0.3:
            modifiers["sovereign_debt_crisis"] = 2.0
        
        # Environmental damage accumulated increases environmental crisis risk
        env_damage = nation_state.get("environmental_damage_accumulated", 0.0)
        if env_damage > 0.6:
            modifiers["environmental_disaster"] = 3.0
        
        # Check each crisis type
        for crisis_key, crisis in CRISIS_TEMPLATES.items():
            prob = base_prob * modifiers.get(crisis_key, 1.0)
            
            # Era-specific adjustments
            if era == "era1_nation_building":
                if crisis_key in ("hyperinflation", "pandemic"):
                    prob *= 0.5  # less common in early 20th c
                if crisis_key == "external_intervention":
                    prob *= 2.0
            elif era == "era2_planned_development":
                if crisis_key == "financial_panic":
                    prob *= 0.3  # planned economy
                if crisis_key == "environmental_disaster":
                    prob *= 1.5  # rapid industrialization
            elif era == "era3_convergence":
                if crisis_key == "financial_panic":
                    prob *= 2.0  # market transition
                if crisis_key == "sovereign_debt_crisis":
                    prob *= 1.5
            
            if self.rng.random() < prob:
                self.active_crises.append(crisis)
                self.crisis_history.append({
                    "year": year,
                    "crisis": crisis_key,
                    "severity": crisis.severity,
                })
                return crisis
        
        return None
    
    def check_random_event(self, year: int) -> Optional[Event]:
        """Check for random non-crisis events."""
        for event_key, event in EVENT_TEMPLATES.items():
            if self.rng.random() < event.probability_per_year:
                self.event_history.append({
                    "year": year,
                    "event": event_key,
                })
                return event
        return None
    
    def cascade_check(self) -> List[Crisis]:
        """Check if any active crises trigger cascading crises."""
        new_crises = []
        for active in self.active_crises:
            for trigger_key in active.can_trigger:
                if trigger_key not in CRISIS_TEMPLATES:
                    continue  # skip unknown cascade triggers
                if self.rng.random() < 0.12:  # 12% cascade chance (was 30%)
                    cascade = CRISIS_TEMPLATES[trigger_key]
                    if cascade not in self.active_crises:
                        new_crises.append(cascade)
                        self.active_crises.append(cascade)
        return new_crises
    
    def resolve_crises(self, year: int) -> List[Crisis]:
        """Decrement crisis durations and resolve expired ones."""
        resolved = []
        surviving = []
        for crisis in self.active_crises:
            crisis.duration_years -= 1
            if crisis.duration_years <= 0:
                resolved.append(crisis)
            else:
                surviving.append(crisis)
        self.active_crises = surviving
        return resolved
    
    @property
    def reform_window_multiplier(self) -> float:
        """How much do active crises boost reform capacity?
        
        Multiple crises stack but with diminishing returns.
        """
        if not self.active_crises:
            return 1.0
        
        total_bonus = sum(c.reform_window_bonus for c in self.active_crises)
        # Diminishing returns: sqrt so 2 crises = 1.4x not 2x
        return 1.0 + math.sqrt(total_bonus) * 0.5
    
    @property
    def total_crisis_impact(self) -> Dict[str, float]:
        """Aggregate all active crisis effects on the nation."""
        impact = {
            "gdp_impact_pct": 0.0,
            "institutional_quality_impact": 0.0,
            "social_cohesion_impact": 0.0,
            "d_coefficient_impact": 0.0,
        }
        
        for crisis in self.active_crises:
            impact["gdp_impact_pct"] += crisis.gdp_impact_pct
            impact["institutional_quality_impact"] += crisis.institutional_quality_impact
            impact["social_cohesion_impact"] += crisis.social_cohesion_impact
            impact["d_coefficient_impact"] += crisis.d_coefficient_impact
        
        return impact
